### ISSUE TITLE
tui_flush: assertion failed with floats with borders that span width of TUI

### ISSUE THREAD

**Issue by timsofteng** [neovim/neovim#14377]
- `nvim --version`:
NVIM v0.5.0-dev+1233-g82ac44d01

- Operating system/version:
Linux, Debian Testing

- Terminal name/version:
alacritty

- `$TERM`:
tmux-256color

I cought this bug several times by accidentl but unfortunately I can't descibe reproducing steps on clean config.
But I finally find how to reproduce it with my config at least.

On this screencast I use many plugins but bug can be reproduced with only 5 of them:
packer.nvim
vim-obsession
nvim-lspconfig
lsp-saga
vimpeccable

1. Open neovim 0.5 in tmux.
2. Open few splits.
3. Open lsp-saga hover.
4. Try to scroll in popup window.

here is video how it looks:

https://user-images.githubusercontent.com/51912173/115014959-ab016380-9ebb-11eb-913b-66b8682f4e9b.mp4

Outside of tmux or without internal neovim's splits I can't reproduce it.

**Comment by mhinz** [comment]
I have no idea what I'm even seeing there. Is that `:terminal` at the bottom? Or an outside tmux pane?

And then multiple sequences of ASCII are thrown into the terminal and at one point suddenly get executed all at once but are obviously no valid executables or zsh commands.

Do the sequences only get added when you type or automatically? What is happening?

**Comment by mjlbach** [comment]
[This assertion](https://github.com/mjlbach/neovim/blob/7e7ba1e08b0334d019d01b88c2ea803ab015d62f/src/nvim/tui/tui.c#L1238) seems to be failing.

**Comment by timsofteng** [comment]
> I have no idea what I'm even seeing there. Is that `:terminal` at the bottom? Or an outside tmux pane?

It looks like neovim failed and back to terminal pane of tmux. It's definetely not neovim's builtin terminal.

> Do the sequences only get added when you type or automatically? What is happening?
It appears when I typing or click mouse. On the screencast it's some of mouse events I suppose.

**Comment by elianiva** [comment]
FWIW I had a segfault earlier, [this assertion](https://github.com/neovim/neovim/blob/bb7d3790bf08b5519623d261d8235bad77b5c0dd/src/nvim/ui_compositor.c#L368-L368) also failed. I'll try to find a way to reproduce it.

**Comment by HungryJoe** [comment]
- **Nvim version:** `NVIM v0.5.0-dev+1280-gbb7d3790b-dirty` ([diff](https://github.com/neovim/neovim/files/6374006/diff.txt) is one line)
- **Operating system:** `Debian GNU/Linux bullseye/sid`
- **Terminal:** `konsole 20.12.3`
- **$TERM:** xterm-256color
- **valgrind:** valgrind-3.16.1
- **All nvim behavior has been reproduced with `nvim -u NONE`**

It looks like floating windows that exceed the width of the terminal window are considered unscrollable (not sure why, but it's set [here](https://github.com/neovim/neovim/blob/bb7d3790bf08b5519623d261d8235bad77b5c0dd/src/nvim/tui/tui.c#L1148)); thus, they get invalidated when scrolling is attempted, which triggers the assertion @mjlbach identified above.

When the `can_scroll` check is disabled [here](https://github.com/neovim/neovim/blob/bb7d3790bf08b5519623d261d8235bad77b5c0dd/src/nvim/tui/tui.c#L1155), the region is never invalidated, so scrolling is allowed. However, an error arises when running the script @mjlbach linked to with a much smaller width (say, `60`): if the floating window is opened when the terminal's width is brought  above 60, the terminal's width is brought below `60`, the floating window is scrolled, and the terminal's width is again brought above `60`, nvim crashes. The error says `corrupted doubl-linked list`. When run with `valgrind` (no options), I get [this output](https://github.com/neovim/neovim/files/6373958/valgrind_error.log).

Also, for any width, there's a weird visual bug: The left and right border columns on the window's top and bottom text rows turn the color of the larger window's background when scrolling to them, which persists for the duration of the window:

![bottomline](https://user-images.githubusercontent.com/12981482/116023570-faf5ce00-a611-11eb-82b0-b58ba443a1da.png)
![topline](https://user-images.githubusercontent.com/12981482/116023571-faf5ce00-a611-11eb-9c75-a2f0e341f1a2.png)

**tl;dr** It seems like scrolling is disabled for floating windows of the kind generated in @mjlbach 's script; if it's enabled manually, other issues arise.

**Comment by bfredl** [comment]
I think the issue is about borders. the maximum width check is incorrect then (needs to be TUI width - 2)

**Comment by mjlbach** [comment]
Closed by #14483

**Comment by timsofteng** [comment]
Problem still exist but i cant reprosuce steps
![Uploading image.png…]()

**Comment by mjlbach** [comment]
I don't know what that picture is supposed to show, but the assert should be fixed so if you can put into words what the issue is, please open another issue.
